accumulator/batch.go

```diff
@@ -0,0 +1,283 @@
+	}
+	return nil
+}
+
+// OnLambdaLogRuntimeDone prepares the data for the invocation to be shipped
+// to APM Server. It accepts requestID and status of the invocation both of
+// which can be retrieved after parsing `platform.runtimeDone` event.
+func (b *Batch) OnLambdaLogRuntimeDone(reqID, status string) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.finalizeInvocation(reqID, status)
+}
+
+// OnShutdown flushes the data for shipping to APM Server by finalizing all
+// the invocation in the batch. If we haven't received a platform.runtimeDone
+// event for an invocation so far we won't be able to recieve it in time thus
+// the status needs to be guessed based on the available information.
+func (b *Batch) OnShutdown(status string) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	for _, inc := range b.invocations {
+		if err := b.finalizeInvocation(inc.RequestID, status); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddLambdaData adds a new entry to the batch. Returns ErrBatchFull
+// if batch has reached its maximum size.
+func (b *Batch) AddLambdaData(d []byte) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	return b.addData(d)
+}
+
+// Count return the number of APMData entries in batch.
+func (b *Batch) Count() int {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return b.count
+}
+
+// ShouldShip indicates when a batch is ready for sending.
+// A batch is marked as ready for flush when one of the
+// below conditions is reached:
+// 1. max size is greater than threshold (90% of maxSize)
+// 2. batch is older than maturity age
+func (b *Batch) ShouldShip() bool {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return (b.count >= int(float64(b.maxSize)*maxSizeThreshold)) ||
+		(!b.age.IsZero() && time.Since(b.age) > b.maxAge)
+}
+
+// Reset resets the batch to prepare for new set of data
+func (b *Batch) Reset() {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.count, b.age = 0, zeroTime
+	b.buf.Truncate(b.metadataBytes)
+}
+
+// ToAPMData returns APMData with metadata and the accumulated batch
+func (b *Batch) ToAPMData() APMData {
+	b.mu.RLock()
+	defer b.mu.RUnlock()
+	return APMData{
+		Data: b.buf.Bytes(),
+	}
+}
+
+func (b *Batch) finalizeInvocation(reqID, status string) error {
+	inc, ok := b.invocations[reqID]
+	if !ok {
+		return fmt.Errorf("invocation for requestID %s does not exist", reqID)
+	}
+	defer delete(b.invocations, reqID)
+	proxyTxn, err := inc.Finalize(status)
+	if err != nil {
+		return err
+	}
+	return b.addData(proxyTxn)
+}
+
+func (b *Batch) addData(data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+	if b.metadataBytes == 0 {
+		return ErrMetadataUnavailable
+	}
+	if b.count == b.maxSize {
+		return ErrBatchFull
+	}
+	if err := b.buf.WriteByte('\n'); err != nil {
+		return err
+	}
+	if _, err := b.buf.Write(data); err != nil {
+		return err
+	}
+	if b.count == 0 {
+		// For first entry, set the age of the batch
+		b.age = time.Now()
+	}
+	b.count++
+	return nil
+}
+
+func findEventType(body []byte) []byte {
```

Comment:
Should we make this return a string? We're converting the result to a string in all cases where it's called.

Or alternatively, change it to `func isTransactionEvent(body []byte) bool`, to avoid allocating due to conversion from []byte to string.

Reply:
SGTM, refactored the code with `isTransactionEvent`